File: _inc/laravel/app/Http/Controllers/Companies/BranchController.php
<?php

namespace App\Http\Controllers;

use App\Config\Constants\{
    CompaniesConstants,
    DatabaseConstants,
    UsersConstants,
    ViewsConstants,
};
use App\Traits\{ChecksLogin, ChecksPermissions};
use App\Models\{Branch, Department, Employee};
use Illuminate\Http\{JsonResponse, RedirectResponse, Request};
use Illuminate\Support\Facades\{Log, Route, View as ViewFacade};
use Illuminate\View\View;

use function App\Http\Controllers\Helpers\{defaultUndefinedException, defaultPermissionDenial};
use App\Traits\HasCrudConstants;
use App\Traits\DefinesResourceActions;
class BranchController extends Controller
{
	use DefinesResourceActions;

    use HasCrudConstants;

    use ChecksLogin, ChecksPermissions;

    public function index(Request $request): RedirectResponse|View
    {
        $action = __FUNCTION__;
        $method = __METHOD__;
        $class = static::class;
        $base = class_basename($class);
        return $this->measureProfile($action, function () use ($request, $action, $method, $class, $base) {
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
            if (($g = self::guard($request, 'manage branch')) !== true) return $g;
            Log::debug("[{$base}::{$action}] start", [UsersConstants::COL_USER_ID => $request->user()?->id, 'method' => $method]);
            try {
                $qStart = microtime(true);
                $branches = Branch::where(DatabaseConstants::COL_TABLE_CREATOR, $request->user()?->creatorId())->get();
                $this->logExecutionTime($qStart, $action, 'fetchBranches');
                $viewPath = ViewsConstants::BRC . '.' . $action;
                if (!ViewFacade::exists($viewPath)) return back()->with('error', "HTTP 404: Page {$viewPath} not found!");
                $renderStart = microtime(true);
                $resp = view($viewPath, compact('branches'));
                $this->logExecutionTime($renderStart, $action, 'renderView');
                return $resp;
            } catch (\Throwable $e) {
                Log::error("[{$base}::{$action}] error", ['error' => $e->getMessage()]);
                Log::debug("[{$base}::{$action}] exception context", ['exception' => get_class($e), 'file' => $e->getFile(), 'line' => $e->getLine()]);
                return defaultUndefinedException($request, $e, $class . '::' . $action);
            }
        }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $base]);
    }

    public function create(Request $request): RedirectResponse|View
    {
        $action = __FUNCTION__;
        $method = __METHOD__;
        $class = static::class;
        $base = class_basename($class);
        return $this->measureProfile($action, function () use ($request, $action, $method, $base) {
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
            if (($g = self::guard($request, 'create branch')) !== true) return $g;
            Log::debug("[{$base}::{$action}] start", [UsersConstants::COL_USER_ID => $request->user()?->id, 'method' => $method]);
            $viewPath = ViewsConstants::BRC . '.' . $action;
            if (!ViewFacade::exists($viewPath)) return back()->with('error', "HTTP 404: Page {$viewPath} not found!");
            $renderStart = microtime(true);
            $resp = view($viewPath);
            $this->logExecutionTime($renderStart, $action, 'renderView');
            return $resp;
        }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $base]);
    }

    public function store(Request $request): RedirectResponse
    {
        $action = __FUNCTION__;
        $method = __METHOD__;
        $class = static::class;
        $base = class_basename($class);
        return $this->measureProfile($action, function () use ($request, $action, $class, $base) {
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
            if (($g = self::guard($request, 'create branch')) !== true) return $g;
            $valStart = microtime(true);
            $v = validator($request->all(), ['name' => 'required']);
            $this->logExecutionTime($valStart, $action, 'validate');
            if ($v->fails()) return back()->with('error', $v->errors()->first());
            try {
                $crtStart = microtime(true);
                $branch = Branch::create([
                    CompaniesConstants::COL_BRC_NM => $request->input('name'),
                    DatabaseConstants::COL_TABLE_CREATOR => $request->user()?->creatorId(),
                ]);
                $this->logExecutionTime($crtStart, $action, 'createBranch');
                Log::info("[{$base}::{$action}] created", ['branch_id' => $branch->id, UsersConstants::COL_USER_ID => $request->user()?->id]);
                return redirect()->route(ViewsConstants::BRC . '.index')->with('success', __('Branch successfully created.'));
            } catch (\Throwable $e) {
                Log::error("[{$base}::{$action}] error", ['error' => $e->getMessage()]);
                Log::debug("[{$base}::{$action}] exception context", ['exception' => get_class($e), 'file' => $e->getFile(), 'line' => $e->getLine()]);
                return defaultUndefinedException($request, $e, $class . '::' . $action);
            }
        }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $base]);
    }

    public function show(Branch $branch): RedirectResponse
    {
        $action = __FUNCTION__;
        $method = __METHOD__;
        $class = static::class;
        $base = class_basename($class);
        return $this->measureProfile($action, function () use ($branch) {
            return redirect()->route(ViewsConstants::BRC . '.index')->with($branch);
        }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $base]);
    }

    public function edit(Request $request, Branch $branch): RedirectResponse|View
    {
        $action = __FUNCTION__;
        $method = __METHOD__;
        $class = static::class;
        $base = class_basename($class);
        return $this->measureProfile($action, function () use ($request, $branch, $action, $class) {
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
            if (($g = self::guard($request, 'edit branch')) !== true) return $g;
            if ($branch[DatabaseConstants::COL_TABLE_CREATOR] !== $request->user()?->creatorId()) return defaultPermissionDenial($request, null, $class . '::' . $action);
            $viewPath = ViewsConstants::BRC . '.' . $action;
            if (!ViewFacade::exists($viewPath)) return back()->with('error', "HTTP 404: Page {$viewPath} not found!");
            $renderStart = microtime(true);
            $resp = view($viewPath, compact('branch'));
            $this->logExecutionTime($renderStart, $action, 'renderView');
            return $resp;
        }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $base, 'branch_id' => $branch->id]);
    }

    public function update(Request $request, Branch $branch): RedirectResponse
    {
        $action = __FUNCTION__;
        $method = __METHOD__;
        $class = static::class;
        $base = class_basename($class);
        return $this->measureProfile($action, function () use ($request, $branch, $action, $class, $base) {
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
            if (($g = self::guard($request, 'edit branch')) !== true) return $g;
            if ($branch[DatabaseConstants::COL_TABLE_CREATOR] !== $request->user()?->creatorId()) return defaultPermissionDenial($request, null, $class . '::' . $action);
            $valStart = microtime(true);
            $v = validator($request->all(), [CompaniesConstants::COL_BRC_NM => 'required']);
            $this->logExecutionTime($valStart, $action, 'validate');
            if ($v->fails()) return back()->with('error', $v->errors()->first());
            try {
                $updStart = microtime(true);
                $branch->update([CompaniesConstants::COL_BRC_NM => $request->input('name')]);
                $this->logExecutionTime($updStart, $action, 'updateBranch');
                Log::info("[{$base}::{$action}] updated", ['branch_id' => $branch->id, UsersConstants::COL_USER_ID => $request->user()?->id]);
                return redirect()->route(ViewsConstants::BRC . '.index')->with('success', __('Branch successfully updated.'));
            } catch (\Throwable $e) {
                Log::error("[{$base}::{$action}] error", ['error' => $e->getMessage(), 'branch_id' => $branch->id]);
                Log::debug("[{$base}::{$action}] exception context", ['exception' => get_class($e), 'file' => $e->getFile(), 'line' => $e->getLine()]);
                return defaultUndefinedException($request, $e, $class . '::' . $action);
            }
        }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $base, 'branch_id' => $branch->id]);
    }

    public function destroy(Request $request, Branch $branch): RedirectResponse
    {
        $action = __FUNCTION__;
        $method = __METHOD__;
        $class = static::class;
        $base = class_basename($class);
        return $this->measureProfile($action, function () use ($request, $branch, $action, $class, $base) {
            if (($userOrRedirect = self::_checkLogin()) instanceof RedirectResponse) return $userOrRedirect;
            if (($g = self::guard($request, 'delete branch')) !== true) return $g;
            if ($branch[DatabaseConstants::COL_TABLE_CREATOR] !== $request->user()?->creatorId()) return defaultPermissionDenial($request, null, $class . '::' . $action);
            try {
                $delStart = microtime(true);
                $branch->delete();
                $this->logExecutionTime($delStart, $action, 'deleteBranch');
                Log::info("[{$base}::{$action}] deleted", ['branch_id' => $branch->id, UsersConstants::COL_USER_ID => $request->user()?->id]);
                return redirect()->route(ViewsConstants::BRC . '.index')->with('success', __('Branch successfully deleted.'));
            } catch (\Throwable $e) {
                Log::error("[{$base}::{$action}] error", ['error' => $e->getMessage(), 'branch_id' => $branch->id]);
                Log::debug("[{$base}::{$action}] exception context", ['exception' => get_class($e), 'file' => $e->getFile(), 'line' => $e->getLine()]);
                return defaultUndefinedException($request, $e, $class . '::' . $action);
            }
        }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $base, 'branch_id' => $branch->id]);
    }

    public const GET_DPT = 'getDepartment';
    public function getDepartment(Request $request): JsonResponse
    {
        $action = __FUNCTION__;
        $method = __METHOD__;
        $class = static::class;
        $base = class_basename($class);
        return $this->measureProfile($action, function () use ($request, $action) {
            if ((self::_checkLogin()) instanceof RedirectResponse) return response()->json([], 401);
            $branchId = (int) $request->input(CompaniesConstants::COL_BRC_ID, 0);
            $qStart = microtime(true);
            $departments = $branchId === 0
                ? Department::pluck(CompaniesConstants::COL_DEP_NM, 'id')->toArray()
                : Department::where(CompaniesConstants::COL_BRC_ID, $branchId)->pluck(CompaniesConstants::COL_DEP_NM, 'id')->toArray();
            $this->logExecutionTime($qStart, $action, 'fetchDepartments');
            return response()->json($departments);
        }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $base, 'branch_id' => $request->input(CompaniesConstants::COL_BRC_ID)]);
    }

    public const GET_EMP = 'getEmployee';
    public function getEmployee(Request $request): JsonResponse
    {
        $action = __FUNCTION__;
        $method = __METHOD__;
        $class = static::class;
        $base = class_basename($class);
        return $this->measureProfile($action, function () use ($request, $action) {
            if ((self::_checkLogin()) instanceof RedirectResponse) return response()->json([], 401);
            $deptIds = $request->input(CompaniesConstants::COL_DEP_ID, []);
            if (!is_array($deptIds)) $deptIds = [];
            $qStart = microtime(true);
            $employees = in_array(0, $deptIds)
                ? Employee::pluck(UsersConstants::COL_NM, 'id')->toArray()
                : Employee::whereIn(CompaniesConstants::COL_DEP_ID, $deptIds)->pluck(UsersConstants::COL_NM, 'id')->toArray();
            $this->logExecutionTime($qStart, $action, 'fetchEmployees');
            return response()->json($employees);
        }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $base, 'dept_ids' => $request->input(CompaniesConstants::COL_DEP_ID)]);
    }
}

File: _inc/_DEPRECATED_django/erp_prestech_backend/app/Http/Controllers/companies/branch_controller.py
import inspect
import logging
from django.contrib import messages
from django.contrib.auth.decorators import login_required
from django.core.exceptions import PermissionDenied
from django.http import HttpRequest, HttpResponse, JsonResponse
from django.shortcuts import get_object_or_404, redirect, render
from django.utils.decorators import method_decorator
from django.views.decorators.http import require_http_methods
from .._helpers.http import get_redirect_url
from .._helpers.error_handlers import default_permission_denial, default_undefined_exception
from .._traits.controller import Controller
from ....Models.companies.branch import Branch
from ....Models.companies.department import Department
from ....Models.individuals.employee import Employee
logger = logging.getLogger(__name__)

class BranchController(Controller):

    @method_decorator(login_required)
    @classmethod
    def index(cls, request: HttpRequest) -> HttpResponse:
        CN, MN = cls.__name__, inspect.currentframe().f_code.co_name
        try:
            if not request.user.has_perm('crm.manage_branch'):
                return default_permission_denial(
                    request,
                    err=PermissionDenied('User lacks permission: crm.manage_branch'),
                    ref=f'{CN}::{MN}',
                    logger=logger
                )
            creator_id = request.user.creator_id
            branches   = Branch.objects.filter(created_by=creator_id)
            return render(request, 'branch/index.html', {'branches': branches})
        except Exception as e:
            return default_undefined_exception(request, err=e, ref=f'{CN}::{MN}', logger=logger)

    @method_decorator(login_required)
    @classmethod
    def create(cls, request: HttpRequest) -> HttpResponse:
        CN, MN = cls.__name__, inspect.currentframe().f_code.co_name
        try:
            if not request.user.has_perm('crm.create_branch'):
                return default_permission_denial(
                    request,
                    err=PermissionDenied('User lacks permission: crm.create_branch'),
                    ref=f'{CN}::{MN}',
                    logger=logger
                )
            return render(request, 'branch/create.html')
        except Exception as e:
            return default_undefined_exception(request, err=e, ref=f'{CN}::{MN}', logger=logger)

    @require_http_methods(["POST"])
    @method_decorator(login_required)
    @classmethod
    def store(cls, request: HttpRequest) -> HttpResponse:
        CN, MN = cls.__name__, inspect.currentframe().f_code.co_name
        try:
            if not request.user.has_perm('crm.create_branch'):
                return default_permission_denial(
                    request,
                    err=PermissionDenied('User lacks permission: crm.create_branch'),
                    ref=f'{CN}::{MN}',
                    logger=logger
                )
            name = request.POST.get('name')
            if not name:
                messages.error(request, "Name is required.")
                return redirect(get_redirect_url(request))
            branch = Branch(name=name, created_by=request.user.creator_id)
            branch.save()
            messages.success(request, "Branch successfully created.")
            return redirect('branch_index')
        except Exception as e:
            return default_undefined_exception(request, err=e, ref=f'{CN}::{MN}', logger=logger)

    @method_decorator(login_required)
    @classmethod
    def show(cls, request: HttpRequest, branch_id: int) -> HttpResponse:
        # CN, MN = cls.__name__, inspect.currentframe().f_code.co_name
        # TODO no actual show logic; redirect to index
        return redirect('branch_index')

    @method_decorator(login_required)
    @classmethod
    def edit(cls, request: HttpRequest, branch_id: int) -> HttpResponse:
        CN, MN = cls.__name__, inspect.currentframe().f_code.co_name
        try:
            if not request.user.has_perm('crm.edit_branch'):
                return JsonResponse({'error': "Permission denied."}, status=401)
            branch = get_object_or_404(Branch, pk=branch_id)
            if branch.created_by != request.user.creator_id:
                return JsonResponse({'error': "Permission denied."}, status=401)
            return render(request, 'branch/edit.html', {'branch': branch})
        except Exception as e:
            return default_undefined_exception(request, err=e, ref=f'{CN}::{MN}', logger=logger)

    @require_http_methods(["POST"])
    @method_decorator(login_required)
    @classmethod
    def update(cls, request: HttpRequest, branch_id: int) -> HttpResponse:
        CN, MN = cls.__name__, inspect.currentframe().f_code.co_name
        try:
            if not request.user.has_perm('crm.edit_branch'):
                return default_permission_denial(
                    request,
                    err=PermissionDenied('User lacks permission: crm.edit_branch'),
                    ref=f'{CN}::{MN}',
                    logger=logger
                )
            branch = get_object_or_404(Branch, pk=branch_id)
            if branch.created_by != request.user.creator_id:
                return default_permission_denial(
                    request,
                    err=PermissionDenied('Permission denied.'),
                    ref=f'{CN}::{MN}',
                    logger=logger
                )
            name = request.POST.get('name')
            if not name:
                messages.error(request, "Name is required.")
                return redirect(get_redirect_url(request))
            branch.name = name
            branch.save()
            messages.success(request, "Branch successfully updated.")
            return redirect('branch_index')
        except Exception as e:
            return default_undefined_exception(request, err=e, ref=f'{CN}::{MN}', logger=logger)

    @require_http_methods(["POST"])
    @method_decorator(login_required)
    @classmethod
    def destroy(cls, request: HttpRequest, branch_id: int) -> HttpResponse:
        CN, MN = cls.__name__, inspect.currentframe().f_code.co_name
        try:
            if not request.user.has_perm('crm.delete_branch'):
                return default_permission_denial(
                    request,
                    err=PermissionDenied('User lacks permission: crm.delete_branch'),
                    ref=f'{CN}::{MN}',
                    logger=logger
                )
            branch = get_object_or_404(Branch, pk=branch_id)
            if branch.created_by != request.user.creator_id:
                return default_permission_denial(
                    request,
                    err=PermissionDenied('Permission denied.'),
                    ref=f'{CN}::{MN}',
                    logger=logger
                )
            branch.delete()
            messages.success(request, "Branch successfully deleted.")
            return redirect('branch_index')
        except Exception as e:
            return default_undefined_exception(request, err=e, ref=f'{CN}::{MN}', logger=logger)

    @method_decorator(login_required)
    @classmethod
    def get_department(cls, request: HttpRequest) -> JsonResponse:
        CN, MN = cls.__name__, inspect.currentframe().f_code.co_name
        try:
            branch_id = request.GET.get('branch_id')
            qs = Department.objects.all().values('id', 'name') \
                 if branch_id in (None, '0') \
                 else Department.objects.filter(branch_id=branch_id).values('id', 'name')
            department_dict = {d['id']: d['name'] for d in qs}
            return JsonResponse(department_dict)
        except Exception as e:
            return default_undefined_exception(request, err=e, ref=f'{CN}::{MN}', logger=logger)

    @method_decorator(login_required)
    @classmethod
    def get_employee(cls, request: HttpRequest) -> JsonResponse:
        CN, MN = cls.__name__, inspect.currentframe().f_code.co_name
        try:
            dept_ids = request.GET.getlist('department_id')
            qs = Employee.objects.all().values('id', 'name') \
                 if '0' in dept_ids \
                 else Employee.objects.filter(department_id__in=dept_ids).values('id', 'name')
            employee_dict = {e['id']: e['name'] for e in qs}
            return JsonResponse(employee_dict)
        except Exception as e:
            return default_undefined_exception(request, err=e, ref=f'{CN}::{MN}', logger=logger)
